lib/Lower/Mangler.cpp

```diff
@@ -62,15 +67,28 @@ Fortran::lower::mangle::mangleName(fir::NameUniquer &uniquer,
             return uniquer.doProgramEntry().str();
           },
           [&](const Fortran::semantics::SubprogramDetails &) {
-            auto &cb{symbol->name()};
-            auto modNames{moduleNames(symbol->scope())};
-            return uniquer.doProcedure(modNames, hostName(symbol->scope()),
-                                       toStringRef(cb));
+            const auto *scope{&symbol->owner()};
+            auto &symbolName{symbol->name()};
+            // Separate module subprograms must be mangled according to the
+            // scope where they were declared (the symbol we have is the
+            // definition).
+            if (symbol->attrs().test(Fortran::semantics::Attr::MODULE) &&
+                scope->IsSubmodule()) {
+              // FIXME symbol from MpSubprogramStmt do not seem to have
+              // Attr::MODULE set.
```

Comment:
Is this caught by the following `assert`?

Reply:
Sadly no, it gets unnoticed because there seems to be no way here to tell that this symbol comes from a `parser::MpSubprogramStmt` vs a normal subprogram, so it is mangled as a normal subprogram which is wrong.
I've asked Tim is there is a better way to detect this.